osquery/tables/system/darwin/smc_keys.cpp

```diff
@@ -379,28 +382,99 @@ inline uint32_t strtoul(const char *str, size_t size, size_t base) {
+  uint64_t total = 0;
+  for (size_t i = 0; i < size; i++) {
+    total += static_cast<uint64_t>(static_cast<unsigned char>(str[i]))
+             << (size - 1 - i) * 8;
+  }
+  return total;
+}
+
+double getConvertedValue(const std::string& smcType,
+                         const std::string& smcVal) {
+  // Some types are returned w/ whitespace at the end.
+  std::string type = boost::trim_right_copy(smcType);
+
   std::string val;
   try {
     val = boost::algorithm::unhex(smcVal);
   } catch (const boost::algorithm::hex_decode_error &e) {
     return -1.0;
   }
 
-  if (val.size() < 2) {
+  int size = val.size();
+
+  if (val.size() < 1) {
+    LOG(ERROR) << "Invalid size 0 for SMC value; type: " << type;
     return -1.0;
   }
-
-  float convertedVal = -1.0;
-  if (smcType == "sp78") {
-    convertedVal = (val[0] * 256 + val[1]) / 256.0;
-  } else if (smcType == "fpe2") {
-    convertedVal = (int(val[0]) << 6) + (int(val[1]) >> 2);
-  } else if (smcType == "sp5a") {
-    convertedVal = (val[0] * 256 + val[1]) / 1024.0;
+  if (val.size() > 8) {
+    LOG(ERROR)
+        << "SMC keys greater than 64 bits are currently unsupported; type: "
+        << type;
+    return -1.0;
   }
 
+  double convertedVal = -1.0;
+
+  // Little-endian 16.16 fixed-point.
+  if (type == "ioft") {
+    std::uint64_t res;
+    std::memcpy(&res, val.data(), sizeof(res));
+    convertedVal = static_cast<double>(res) / 65536.0f;
+  }
+  // Little-endian 32 bit float.
+  else if (type == "flt") {
+    float f;
+    std::memcpy(&f, val.data(), sizeof(f));
+    convertedVal = static_cast<double>(f);
+  }
+  // Big-endian 16 bit fixed/signed points.
+  else if ((type[0] == 'f' || type[0] == 's') && type[1] == 'p') {
+    // The last character represents the # of fractional bits.
+    const char fbc = type[3];
+    const char ibc = type[2];
+    int is_signed = type[0] == 's' ? 1 : 0;
+    int fractionalBits = (fbc >= 'a' ? (fbc - 'a' + 10) : (fbc - '0'));
+    int integerBits = (ibc >= 'a' ? (ibc - 'a' + 10) : (ibc - '0'));
+    // Check that this is a valid type.
+    if (integerBits + fractionalBits + is_signed != 16) {
+      LOG(ERROR) << "Invalid SMC type: " << type << " (integer: " << integerBits
+                 << ", fractional: " << fractionalBits << ")";
+      return -1.0;
+    }
+    int divisor = 1 << fractionalBits;
+    // Get the integer value represented by the decoded hex string, and divide
+    // by the divisor to get the float value.
+    int intVal = strtoul(val.c_str(), size, 16);
+    if (type[0] == 's') {
+      // "sp" types are signed.
+      convertedVal =
+          static_cast<double>(static_cast<int16_t>(intVal)) / divisor;
+    } else {
+      convertedVal = static_cast<double>(intVal) / divisor;
+    }
+  }
+  // Big-endian 8/16/32/64 bit integers.
+  else if ((type[0] == 's' || type[0] == 'u') && type[1] == 'i') {
+    uint64_t intVal = strtoull(val.c_str(), size, 16);
+    if (type[0] == 's') {
```

Comment:
Why combine the `s` and `u` logic if it's just split apart on line 462?

Reply:
It's combining the `si` and `ui` logic, both of which start by interpreting the value as a big-endian int.  What we do with that int depends on whether it's `s`igned or `u`nsigned.

Reply:
Yeah.. I guess I'd duplicate the `strtoull` over the nested `if` But 🤷